docs/proposals/rescheduling.md

```diff
@@ -0,0 +1,496 @@
+decision the scheduler makes. This observation motivates "controlled rescheduling," a
+mechanism by which Kubernetes will "move" already-running pods over time to improve their
+placement. Controlled rescheduling is the subject of this proposal.
+
+Note that the term "move" is not technically accurate -- the mechanism used is that
+Kubernetes will terminate a pod that is managed by a controller, and the controller will
+create a replacement pod that is then scheduled by the pod's scheduler. The terminated
+pod and replacement pod are completely separate pods, and no pod migration is
+implied. However, describing the process as "moving" the pod is approximately accurate
+and easier to understand, so we will use this terminology in the document.
+
+We use the term "rescheduling" to describe any action the system takes to move an
+already-running pod. The decision may be made and executed by any component; we wil
+introduce the concept of a "rescheduler" component later, but it is not the only
+component that can do rescheduling.
+
+This proposal primarily focuses on the architecture and features/mechanisms used to
+achieve rescheduling, and only briefly discuss example policies. We expect that community
+experimentation will lead to a significantly better understanding of the range, potential,
+and limitations of rescheduling policies.
+
+## Example use cases
+
+Example use cases for rescheduling are
+* moving a running pod onto a node that better satisfies its scheduling criteria
+  * moving a pod onto an under-utilized node
+  * moving a pod onto a node that meets more of the pod's affinity/anti-affinity preferences
+* moving a running pod off of a node in anticipation of a known or speculated future event
+  * draining a node in preparation for maintenance, decomissioning, auto-scale-down, etc.
+  * "preempting" a running pod to make room for a pending pod to schedule
+  * proactively/speculatively make room for large and/or exclusive pods to facilitate
+    fast scheduling in the future (often called "defragmentation")
+  * (note that these last two cases are the only use cases where the first-order intent
+    is to move a pod specifically for the benefit of another pod)
+* moving a running pod off of a node from which it is receiving poor service
+  * anomalous crashlooping or other mysterious incompatiblity between the pod and the node
+  * repeated out-of-resource killing (see #18724)
+  * repeated attempts by the scheduler to schedule the pod onto some node, but it is
+    rejected by Kubelet admission control due to incomplete scheduler knowledge
+  * poor performance due to interference from other containers on the node (CPU hogs,
+    cache thrashers, etc.) (note that in this case there is a choice of moving the victim
+    or the aggressor)
+
+## Some axes of the design space
+
+Among the key design decisions are
+* how does a pod specify its tolerance for these system-generated disruptions, and how
```

Comment:
See the discussion later about disruption budgets. The question you're asking is what is the granularity covered by a single disruption budget. Probably for a first cut, we'd want service granularity for long-running services, and Job granularity for run-to-completion. The reason ReplicaSet is not good is because we create a new ReplicaSet when we do rolling update. Of course we can point the new ReplicaSet at the same disruption budget, but in that case you're really saying the granularity is logically something higher-level than the controller.